pkg/dnsmasq/deployment.go

```diff
@@ -0,0 +1,192 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dnsmasq
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	networkv1 "github.com/openstack-k8s-operators/infra-operator/apis/network/v1beta1"
+
+	common "github.com/openstack-k8s-operators/lib-common/modules/common"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/affinity"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/env"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	// ServiceCommand -
+	ServiceCommand = "dnsmasq"
+	// DebugCommand -
+	DebugCommand = "/bin/sleep infinity"
+)
+
+// Deployment func
+func Deployment(
+	instance *networkv1.DNSMasq,
+	configHash string,
+	labels map[string]string,
+	annotations map[string]string,
+	cms *corev1.ConfigMapList,
+) *appsv1.Deployment {
+	runAsUser := int64(0)
+	terminationGracePeriodSeconds := int64(10)
+
+	livenessProbe := &corev1.Probe{
+		// TODO might need tuning
+		TimeoutSeconds:      5,
+		PeriodSeconds:       3,
+		InitialDelaySeconds: 3,
+	}
+	readinessProbe := &corev1.Probe{
+		// TODO might need tuning
+		TimeoutSeconds:      5,
+		PeriodSeconds:       5,
+		InitialDelaySeconds: 5,
+	}
+
+	command := []string{"/bin/bash"}
+	args := []string{"-c"}
+	initArgs := []string{"-c"}
+	if instance.Spec.Debug.Service {
+		args = append(args, DebugCommand)
+		initArgs = append(initArgs, "/bin/true")
+
+		livenessProbe.Exec = &corev1.ExecAction{
+			Command: []string{
+				"/bin/true",
+			},
+		}
+
+		readinessProbe.Exec = &corev1.ExecAction{
+			Command: []string{
+				"/bin/true",
+			},
+		}
+	} else {
+		dnsmasqCmd := []string{ServiceCommand}
+		dnsmasqCmd = append(dnsmasqCmd, "--interface=*")
+		dnsmasqCmd = append(dnsmasqCmd, "--conf-dir=/etc/dnsmasq.d")
+		dnsmasqCmd = append(dnsmasqCmd, "--hostsdir=/etc/dnsmasq.d/hosts")
+		dnsmasqCmd = append(dnsmasqCmd, "--keep-in-foreground")
+		dnsmasqCmd = append(dnsmasqCmd, "--no-daemon")
+		dnsmasqCmd = append(dnsmasqCmd, "--log-debug")
+		dnsmasqCmd = append(dnsmasqCmd, "--bind-interfaces")
+		dnsmasqCmd = append(dnsmasqCmd, "--listen-address=$(POD_IP)")
+		dnsmasqCmd = append(dnsmasqCmd, "--port "+strconv.Itoa(int(DNSPort)))
+		// log to stdout
+		dnsmasqCmd = append(dnsmasqCmd, "--log-facility=-")
+		// dns
+		dnsmasqCmd = append(dnsmasqCmd, "--no-hosts")
+		dnsmasqCmd = append(dnsmasqCmd, "--domain-needed")
+		dnsmasqCmd = append(dnsmasqCmd, "--no-resolv")
```

Comment:
right now you have to configure an upstream dns server via spe.options in the dnsmasq CR. would it be ok to the ocp cluster one set in the resolv.conf?